From 2260fb9d18282ec3dea5dbb2119d0a96580c9d1e Mon Sep 17 00:00:00 2001
From: yvanliang <yvan6liang@outlook.com>
Date: Fri, 9 Oct 2026 15:43:35 +0800
Subject: [PATCH 01/23] GSI Android 17 compatibility:
 0001-property_service-system-override-policy-for-apex-adb-and-logd-props.patch

Change-Id: I8786c44df9a8e547225fa9799877d23258185ef2
---
 init/property_service.cpp | 38 +++++++++++++++++++++++++++++++++++---
 1 file changed, 35 insertions(+), 3 deletions(-)

diff --git a/init/property_service.cpp b/init/property_service.cpp
index 0b843d5..067e046 100644
--- a/init/property_service.cpp
+++ b/init/property_service.cpp
@@ -61,6 +61,7 @@
 #include <selinux/android.h>
 #include <selinux/label.h>
 #include <selinux/selinux.h>
+#include <sys/utsname.h>
 #include <vendorsupport/api_level.h>
 
 #include "debug_ramdisk.h"
@@ -703,6 +704,26 @@ uint32_t InitPropertySet(const std::string& name, const std::string& value) {
 static Result<void> load_properties_from_file(const char*, const char*,
                                               std::map<std::string, std::string>*);
 
+static bool kernel_supports_capex() {
+    //Put a threshold at >= 5.0
+    struct utsname buf;
+    uname(&buf);
+    const char *where = buf.release;
+    int a = atoi(where);
+    if(a >= 5) return true;
+
+    // If there are vendor apexes, we most likely actually need them
+    auto dir = std::unique_ptr<DIR, decltype(&closedir)>{opendir("/vendor/apex"), closedir};
+    if (!dir) {
+        return false;
+    }
+    for (struct dirent* ent = readdir(dir.get()); ent; ent = readdir(dir.get())) {
+        if(strstr(ent->d_name, "apex")) return true;
+    }
+
+    return false;
+}
+
 /*
  * Filter is used to decide which properties to load: NULL loads all keys,
  * "ro.foo.*" is a prefix match, and "ro.foo.bar" is an exact match.
@@ -803,12 +824,23 @@ static void LoadProperties(char* data, const char* filter, const char* filename,
             std::string error;
             if (CheckPermissions(key, *expanded_value, context, cr, &error) == PROP_SUCCESS) {
                 auto it = properties->find(key);
+                const char *new_value = (*expanded_value).c_str();
+
+                if(strcmp("ro.apex.updatable", key) == 0) {
+                    new_value = kernel_supports_capex() ? "true" : "false";
+                }
                 if (it == properties->end()) {
-                    (*properties)[key] = std::move(*expanded_value);
-                } else if (it->second != *expanded_value) {
+                   (*properties)[key] = new_value;
+                } else if (it->second != new_value) {
                     LOG(WARNING) << "Overriding previous property '" << key << "':'" << it->second
                                  << "' with new value '" << *expanded_value << "'";
-                    it->second = std::move(*expanded_value);
+                    if(strcmp("ro.apex.updatable", key) == 0) {
+                        LOG(WARNING) << "... Ignored apex by kernel version";
+                    } else if(strstr(key, "adb") || strstr(key, "secure") || strstr(key, "ro.logd.kernel")) {
+                        LOG(WARNING) << "... Ignored";
+                    } else {
+                        it->second = new_value;
+                    }
                 }
             } else {
                 LOG(ERROR) << "Do not have permissions to set '" << key << "' to '"
